lua/nvim-treesitter/locals.lua

```diff
@@ -34,18 +34,32 @@ function M.get_definition_id(scope, node_text)
   return table.concat({ 'k', node_text or '', scope:range() }, '_')
 end
 
+--- Get definitions of bufnr (unique and sorted by order of appearance).
 function M.get_definitions(bufnr)
   local locals = M.get_locals(bufnr)
 
-  local defs = {}
-
+  -- Make sure the nodes are unique.
+  local nodes_set = {}
   for _, loc in ipairs(locals) do
     if loc.definition then
-      table.insert(defs, loc.definition)
+      M.recurse_local_nodes(loc.definition, function(_, node, _, match)
+        -- lua doesn't compare tables by value,
+        -- use the value from byte count instead.
+        local _, _, start = node:start()
+        nodes_set[start] = {node = node, type = match or ""}
```

Comment:
The issue I'm seeing with this is that it looses any data that gets `#set` with that directive. Might be best to attach all the data that may be attached to the table that contains the node.

Reply:
not sure what you mean, isn't that data in the node itself? Also, we are not using the set directive in locals AFAIK

Reply:
We use it for setting the scope of functions. It's not on the node, it's on the table containing the node.

https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/javascript/locals.scm#L55

Reply:
So prior to this change this would return

```lua
{
  { var = { scope = 'parent', node = node } }
}
```

After this change we would loose some information

```lua
{
  { type = 'var', node = node }
}
```

I think just combining the containing the table with the results of will solve the issue.

Reply:
I see, this is used in https://github.com/nvim-treesitter/nvim-treesitter/blob/79fd22cc5ae24a67eda25eba0f6671cafc0c3e58/lua/nvim-treesitter/locals.lua#L158-L159

I put this as a separate function first https://github.com/nvim-treesitter/nvim-treesitter/pull/468#discussion_r488586023

Maybe that's better/safe? I'm not sure how the other function is handling duplicates

Reply:
The other function, I don't believe, handles any duplicates. Basically just plucks `definitions` off of locals. I think a separate function might be best. I don't think there would be much overlap anyways, since the original get_definitions is pretty trivial.

Reply:
Should that function live here or in the refactor module?

Reply:
I think it could go in either. It basically dedupes, sorts, and groups by type right?

Reply:
It deduplicates and sorts, it doesn't group by type, but it gets the type together with the node

Reply:
I think it's only useful for the refactor mode for now, I'd moved there, we can move it back if it's required for something else.

Reply:
sounds good 👍